rofl-scheduler/src/serverd/auth.rs

```diff
@@ -0,0 +1,229 @@
+use std::sync::{Arc, LazyLock};
+
+use anyhow::{anyhow, Context};
+use axum::{extract, http, response, RequestPartsExt};
+use axum_extra::{
+    headers::{authorization::Bearer, Authorization},
+    TypedHeader,
+};
+use base64::{prelude::BASE64_STANDARD, Engine};
+use chrono::{prelude::*, Duration};
+use rand::RngCore;
+use rustc_hex::FromHex;
+
+use oasis_runtime_sdk::{
+    crypto::signature::{self, PublicKey},
+    types::address::{Address, SignatureAddressSpec},
+};
+
+use super::{error::Error, State};
+
+/// Global JWT keys instance.
+static JWT_KEYS: LazyLock<Keys> = LazyLock::new(Keys::generate);
+
+/// JWT keys.
+struct Keys {
+    encoding: jsonwebtoken::EncodingKey,
+    decoding: jsonwebtoken::DecodingKey,
+}
+
+impl Keys {
+    /// Generate a new pair of JWT keys.
+    fn generate() -> Self {
+        let mut key = [0; 32];
+        rand::rngs::OsRng.fill_bytes(&mut key);
+
+        Self {
+            encoding: jsonwebtoken::EncodingKey::from_secret(&key),
+            decoding: jsonwebtoken::DecodingKey::from_secret(&key),
+        }
+    }
+}
+
+/// Claims issued after successful authentication.
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct Claims {
+    pub address: String,
+    exp: u64,
+}
+
+impl<S> extract::FromRequestParts<S> for Claims
+where
+    S: Send + Sync,
+{
+    type Rejection = Error;
+
+    async fn from_request_parts(
+        parts: &mut http::request::Parts,
+        _state: &S,
+    ) -> Result<Self, Self::Rejection> {
+        // Extract the token from the authorization header.
+        let TypedHeader(Authorization(bearer)) = parts
+            .extract::<TypedHeader<Authorization<Bearer>>>()
+            .await
+            .map_err(|_| Error::BadAuthToken)?;
+
+        // Decode the user data.
+        let token_data = jsonwebtoken::decode::<Claims>(
+            bearer.token(),
+            &JWT_KEYS.decoding,
+            &jsonwebtoken::Validation::default(),
+        )
+        .map_err(|_| Error::BadAuthToken)?;
+
+        Ok(token_data.claims)
+    }
+}
+
+/// Login request.
+#[derive(Debug, Clone, serde::Deserialize)]
+#[serde(tag = "method", content = "data")]
+pub enum AuthLoginRequest {
+    #[serde(rename = "siwe")]
+    Siwe { message: String, signature: String },
+
+    #[serde(rename = "std")]
+    Std { body: String, signature: String },
+}
+
+impl AuthLoginRequest {
+    /// Verify the login request and return the authenticated claims.
+    pub async fn verify(&self, domain: &str, provider: Address) -> Result<Claims, anyhow::Error> {
+        match self {
+            Self::Siwe { message, signature } => {
```

Comment:
SIWE also contains other useful fields we could enforce for security:
- chain ID (sapphire?)
- contract address (hex-encoded address of the app ID?)
- statement (some meaningful text shown in the Metamask UI when signing)
- resources (also arbitrary text, can be the scheduler endpoint path e.g. `/rofl-scheduler/v1/logs/get`)

Reply:
These seem quite useless for this use case, binding to the domain is IMO enough. The non-SIWE variant also has the provider address which we could add somewhere.

Reply:
Will add the required statement of `Provider: oasis1...` to bind the message to a specific provider. Although the domain is enough in theory.